hazelcast/src/main/java/com/hazelcast/ringbuffer/Ringbuffer.java

```diff
@@ -0,0 +1,156 @@
+     * the actual storage of the item since only during storage the sequenceId is determined. This coupling prevents
+     * certain future optimizations like transparent write-batching: so waiting till a certain number of items
+     * are added or a buffer timeout has expired.
+     *
+     * @param item the item to add.
+     * @throws java.lang.NullPointerException if item is null.
+     */
+    void add(E item);
+
+    /**
+     * Reads an item from the Ringbuffer.
+     *
+     * This method is not destructive unlike e.g. a queue.take. So the same item can be read by multiple readers or it can be
+     * read multiple times by the same reader.
+     *
+     * @param sequenceId the sequenceId of the item to read.
+     * @return the read item
+     * @throws StaleSequenceIdException           if the sequenceId is smaller then {@link #headSequenceId()}. Because a
+     *                                            Ringbuffer won't store all event indefinitely, it can be that the data for the
+     *                                            given sequenceId doesn't exist anymore and the {@link StaleSequenceIdException}
+     *                                            is thrown. It is up to the caller to deal with this particular situation, e.g.
+     *                                            throw an Exception or restart from the last known head. That is why the
+     *                                            StaleSequenceException contains the last known head.
+     * @throws java.lang.IllegalArgumentException if sequenceId is smaller than 0 or larger than {@link #tailSequenceId()}
+     */
+    E read(long sequenceId);
+
+    /**
+     * Adds all the items of a collection to the tail of the Ringbuffer.
+     *
+     * A addAll is likely to outperform multiple calls to {@link #add(Object)} due to better io utilization and a reduced number
+     * of executed operations.
+     *
+     * If the batch is empty, the call is ignored.
+     *
+     * When the collection is not empty, the content is copied into a different data-structure. This means that:
+     * <ol>
+     * <li>after this call completes, the collection can be re-used.</li>
+     * <li>the collection doesn't need to be serializable</li>
+     * </ol>
+     *
+     * If the collection is larger than the capacity of the ringbuffer, then the items that were written first will
+     * be overwritten. Therefor this call will not block till until space comes available.
+     *
+     * The items are inserted in the order of the Iterator of the collection. If an addAll is executed concurrently with
+     * an add or addAll, no guarantees are given that items are contiguous.
+     *
+     * @param collection the batch of items to add.
+     * @throws java.lang.NullPointerException if batch is null, or if an item in this batch is null.
+     */
+    ICompletableFuture addAllAsync(Collection<? extends E> collection);
```

Comment:
Nope. I wanted to make the type Void. but that doesn't work currently with our InvocationFutures.